MarsRover/RoverClient.cs

```diff
@@ -0,0 +1,92 @@
+
+
+        public RoverClient()
+        {
+            _grid = new Grid(3, 3, 5, 5);
+            _heading = new NorthHeading();
+            _obstacles = CreateObstacles(3);
+            _rover = new Rover(_grid, _heading, _obstacles);
+            AddToNavigationHistory();
+        }
+
+        
+        public void RoverInstructions()
+        {
+            var commands = string.Empty;
+            DisplayInstructions();
+            
+            while(!commands.Equals("X"))
+            {
+                Console.WriteLine($"Your current position on the surface grid is {_rover.XCoordinate()}, {_rover.YCoodinate()} and the Rover is facing {_rover.Direction()}.\n");
+                Console.WriteLine("Please enter your rover commands: ");
+                commands = Console.ReadLine().ToUpper().Replace(" ", "");
+                _rover = _rover.ExecuteCommands(commands);
+
+                AddToNavigationHistory();
+                if (_rover.ObstacleDetected()) { ObstacleAvoidance(); } 
+            }
+        }
+
+
+        public List<Obstacle> CreateObstacles(int numberOfObstacles)
+        {
+            var obstacles = new List<Obstacle>();
+            var random = new Random();
+
+            for (var i = 0; i < numberOfObstacles; i++)
+            {
+                var obstacleXCoordinate = random.Next(0, _grid.GridWidth());
+                var obstacleYCoordinate = random.Next(0, _grid.GridHeight());
+
+                if (!(ObstacleSameAsStartingPoint(obstacleXCoordinate, obstacleYCoordinate)))
+                {
+                    var obstacle = new Obstacle(obstacleXCoordinate, obstacleYCoordinate);
+                    obstacles.Add(obstacle);
+                }
+            }
+            return obstacles;
+        }
+
+
+        public bool ObstacleSameAsStartingPoint(int x, int y) => (x == _grid.XCoordinate() && y == _grid.YCoordinate());
+
+
+        private void DisplayInstructions()
+        {
+            Console.WriteLine("****Mars Rover Controller****\n");
+            Console.WriteLine("Use the following letters to write a set of commands for the Rover to follow.  Invalid commands will be ignored:");
+            Console.WriteLine("         F = Forward     B = Backward");
+            Console.WriteLine("         L = Turn Left   R = Turn Right");
+            Console.WriteLine("            X = Exit the Controller\n");
+        }
+
+
+        private void ObstacleAvoidance()
+        {   
+            Console.WriteLine($"An obstacle was detected while trying to perform a move command to {_rover.XCoordinate()}, {_rover.YCoodinate()}.");
+            Console.WriteLine("Rover movement was stopped at the grid location prior to the obstacle detection.\n");
+            _rover = _navigationHistory[_navigationHistory.Count - 2].rover;
+        }
+
+
+        public void AddToNavigationHistory()
+        {
+            var navigation = new RoverNavigationHistory();
```

Comment:
I would challenge the name of `RoverNavigationHistory` here.  I would argue the way this is being used `History` is a misnomer...it's really a "snapshot" into the history.  A collection of the snapshots _is_ the history, but a single item isn't the history if that makes sense.

Reply:
`RoverNavigationHistory`  function was removed